test-data/unit/check-classes.test

```diff
@@ -894,6 +892,280 @@ a.f = a.f # E: Property "f" defined in "A" is read-only
 a.f.x # E: "int" has no attribute "x"
 [builtins fixtures/property.pyi]
 
+-- Descriptors
+-- -----------
+
+
+[case testAccessingNonDataDescriptor]
+class D:
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+class A:
+    f = D()
+
+a = A()
+reveal_type(a.f)  # E: Revealed type is 'builtins.str'
+
+[case testSettingDataDescriptor]
+class D:
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()
+
+a = A()
+a.f = ''
+a.f = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testReadingDescriptorWithoutDunderGet]
+from typing import Union
+
+class D:
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()
+
+    def __init__(self):
+        self.f = 'some string'
+
+a = A()
+reveal_type(a.f)  # E: Revealed type is '__main__.D'
+
+[case testAccessingGenericNonDataDescriptor]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class D(Generic[V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
+        return self.value
+
+class A:
+    f = D(10)
+    g = D('10')
+
+a = A()
+reveal_type(a.f)  # E: Revealed type is 'builtins.int*'
+reveal_type(a.g)  # E: Revealed type is 'builtins.str*'
+
+[case testSettingGenericDataDescriptor]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class D(Generic[V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
+        return self.value
+    def __set__(self, instance: T, value: V) -> None:
+        pass
+
+class A:
+    f = D(10)
+    g = D('10')
+
+a = A()
+a.f = 1
+a.f = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+a.g = ''
+a.g = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testAccessingGenericDescriptorFromClass]
+# flags: --strict-optional
+from d import D
+
+class A:
+    f = D(10)
+    g = D('10')
+
+reveal_type(A.f)  # E: Revealed type is '__main__.D[builtins.int*]'
+reveal_type(A.g)  # E: Revealed type is '__main__.D[builtins.str*]'
+[file d.pyi]
+from typing import TypeVar, Type, Generic, overload
+
+T = TypeVar('T')
+V = TypeVar('V')
+class D(Generic[V]):
+    def __init__(self, value: V) -> None: pass
+
+    @overload
+    def __get__(self, instance: None, owner: Type[T]) -> 'D[V]': pass
+    @overload
+    def __get__(self, instance: T, owner: Type[T]) -> V: pass
+
+[file d.py]
```

Comment:
I think not, it'll just confuse people who know how the test system works. The test system also doesn't run the tests at all, so what they do at runtime is neither tested nor relevant.